lib/core/src/persist/sync.rs

```diff
@@ -0,0 +1,640 @@
+    ) -> Result<()> {
+        let record_id = Record::get_id_from_record_type(record_type, data_id);
+        let updated_fields = updated_fields
+            .map(|fields| {
+                let fields = fields
+                    .iter()
+                    .map(|field| format!("'$[#]', '{field}'"))
+                    .collect::<Vec<String>>()
+                    .join(",");
+                format!("json_insert(
+                    COALESCE((SELECT updated_fields_json FROM sync_outgoing WHERE record_id = :record_id), '[]'),
+                    {fields}
+                )")
+            })
+            .unwrap_or("NULL".to_string());
+        con.execute(&format!("
+            INSERT OR REPLACE INTO sync_outgoing(record_id, data_id, record_type, commit_time, updated_fields_json)
+            VALUES(
+                :record_id, 
+                :data_id, 
+                :record_type, 
+                :commit_time, 
+                {updated_fields}
+            ) 
+        "),
+            named_params! {
+                ":record_id": record_id,
+                ":data_id": data_id,
+                ":record_type": record_type,
+                ":commit_time": utils::now(),
+            },
+        )?;
+
+        Ok(())
+    }
+
+    fn select_sync_outgoing_changes_query(where_clauses: Vec<String>) -> String {
+        let mut where_clause_str = String::new();
+        if !where_clauses.is_empty() {
+            where_clause_str = String::from("WHERE ");
+            where_clause_str.push_str(where_clauses.join(" AND ").as_str());
+        }
+
+        format!(
+            "
+            SELECT 
+                record_id,
+                data_id,
+                record_type,
+                commit_time,
+                updated_fields_json
+            FROM sync_outgoing
+            {where_clause_str}
+        "
+        )
+    }
+
+    fn sql_row_to_sync_outgoing_changes(row: &Row) -> Result<SyncOutgoingChanges> {
+        let record_id = row.get(0)?;
+        let data_id = row.get(1)?;
+        let record_type = row.get(2)?;
+        let commit_time = row.get(3)?;
+        let updated_fields = match row.get::<_, Option<String>>(4)? {
+            Some(fields) => Some(serde_json::from_str(&fields)?),
+            None => None,
+        };
+
+        Ok(SyncOutgoingChanges {
+            record_id,
+            data_id,
+            record_type,
+            commit_time,
+            updated_fields,
+        })
+    }
+
+    pub(crate) fn get_sync_outgoing_changes(&self) -> Result<Vec<SyncOutgoingChanges>> {
+        let con = self.get_connection()?;
+
+        let query = Self::select_sync_outgoing_changes_query(vec![]);
+        let mut stmt = con.prepare(&query)?;
+        let mut rows = stmt.query([])?;
+
+        let mut outgoing_changes = vec![];
+        while let Some(row) = rows.next()? {
+            let detail = Self::sql_row_to_sync_outgoing_changes(row)?;
+            outgoing_changes.push(detail);
+        }
```

Comment:
The problem is query_map and query_row (like below) expect a `rusqlite::Result`, whereas we're converting into an `anyhow::Result` since the mapping logic also does JSON deserialization (of the `updated_fields_json` column).